lib/ecto/repo/schema.ex

```diff
@@ -573,7 +574,17 @@ defmodule Ecto.Repo.Schema do
   end
 
   defp replace_all_fields!(_kind, schema) do
-    schema.__schema__(:fields)
+    Enum.map(schema.__schema__(:fields), &field_source!(schema, &1))
+  end
+
+  defp field_source!(schema, field) do
+    case schema.__schema__(:field_source, field) do
+      nil ->
+        raise ArgumentError, "unknown field for :on_conflict, got: #{inspect(field)}"
```

Comment:
I'm not sure if this is desired but I think it would be helpful in case users do something like:

```ex
MyRepo.insert!(%MySchema{foo: :foo}, on_confict: {:replace, [:unknown]}, ...)
```

Instead of raising on `nil` field (returned by `__schema__(:field_source, field)`) not existing in lower layers, this raises more meaningful error IMO.